File: csharp/DataStructures/src/dsa/tree/BinarySearchTree.cs
namespace dsa.tree;

public class BinarySearchTree
{
    // static void Main()
    // {
    //     BinarySearchTree bst = new();
    //     bst.insert(5);
    //     bst.insert(3);
    //     bst.insert(7);
    //     bst.insert(1);

    //     bst.inOrder();

    //     Console.WriteLine();

    //     if (null != bst.search(10))
    //         Console.WriteLine("Key found");
    //     else
    //         Console.WriteLine("Key not found");
    // }
    
    // A raiz da árvore
    public TreeNode root;

    // Classe interna para representar um nó da árvore
    public class TreeNode
    {
        // O valor do nó
        public int data;
        // O nó filho à esquerda
        public TreeNode left;
        // O nó filho à direita
        public TreeNode right;

        // Construtor do nó que inicializa o valor do nó
        public TreeNode(int data) => this.data = data;
    }

    // Método para inserir um valor na árvore
    public void insert(int value) => root = insert(root, value);

    // Método recursivo para inserir um valor na árvore
    public TreeNode insert(TreeNode root, int value)
    {
        // Se a árvore estiver vazia, cria um novo nó
        if (root == null)
        {
            root = new TreeNode(value);
            return root;
        }

        // Se o valor for menor que o valor do nó atual, insere à esquerda
        if (value < root.data)
            root.left = insert(root.left, value);
        // Se o valor for maior que o valor do nó atual, insere à direita
        else
            root.right = insert(root.right, value);

        return root;
    }

    // Método para imprimir os valores da árvore em ordem
    public void inOrder() => inOrder(root);

    // Método recursivo para imprimir os valores da árvore em ordem
    public void inOrder(TreeNode root)
    {
        // Se a árvore estiver vazia, retorna
        if (root == null)
            return;

        // Primeiro imprime o valor do nó à esquerda
        inOrder(root.left);
        // Depois imprime o valor do nó atual
        Console.Write(root.data + " ");
        // Por fim, imprime o valor do nó à direita
        inOrder(root.right);
    }

    // Método para procurar um valor na árvore
    public TreeNode search(int key) => search(root, key);

    // Método recursivo para procurar um valor na árvore
    public TreeNode search(TreeNode root, int key)
    {
        // Se a árvore estiver vazia ou se o valor do nó atual for o valor procurado, retorna o nó
        if (root == null || root.data == key)
            return root;

        // Se o valor procurado for menor que o valor do nó atual, procura à esquerda
        if (key < root.data)
            return search(root.left, key);
        // Se o valor procurado for maior que o valor do nó atual, procura à direita
        else
            return search(root.right, key);
    }
}

File: go/dsa/tree/BinarySearchTree/BinarySearchTree.go
package binarysearchtree

import "fmt"

// BinarySearchTree é a estrutura principal da árvore binária de busca
type BinarySearchTree struct {
	root *TreeNode // A raiz da árvore
}

// TreeNode representa um nó da árvore
type TreeNode struct {
	data  int       // O valor do nó
	left  *TreeNode // O nó filho à esquerda
	right *TreeNode // O nó filho à direita
}

// NewBinarySearchTree cria e retorna uma nova instância de BinarySearchTree
func NewBinarySearchTree() *BinarySearchTree {
	return &BinarySearchTree{}
}

// Insert insere um valor na árvore
func (bst *BinarySearchTree) Insert(value int) {
	bst.root = bst.insertNode(bst.root, value)
}

// insertNode é um método recursivo para inserir um valor na árvore
func (bst *BinarySearchTree) insertNode(root *TreeNode, value int) *TreeNode {
	// Se a árvore estiver vazia, cria um novo nó
	if root == nil {
		return &TreeNode{data: value}
	}
	// Se o valor for menor que o valor do nó atual, insere à esquerda
	if value < root.data {
		root.left = bst.insertNode(root.left, value)
	} else { // Se o valor for maior que o valor do nó atual, insere à direita
		root.right = bst.insertNode(root.right, value)
	}
	return root
}

// InOrder imprime os valores da árvore em ordem
func (bst *BinarySearchTree) InOrder() {
	bst.inOrderTraversal(bst.root)
	fmt.Println()
}

// inOrderTraversal é um método recursivo para imprimir os valores da árvore em ordem
func (bst *BinarySearchTree) inOrderTraversal(root *TreeNode) {
	if root == nil {
		return
	}
	// Primeiro imprime o valor do nó à esquerda
	bst.inOrderTraversal(root.left)
	// Depois imprime o valor do nó atual
	fmt.Print(root.data, " ")
	// Por fim, imprime o valor do nó à direita
	bst.inOrderTraversal(root.right)
}

// Search procura um valor na árvore
func (bst *BinarySearchTree) Search(key int) *TreeNode {
	return bst.searchNode(bst.root, key)
}

// searchNode é um método recursivo para procurar um valor na árvore
func (bst *BinarySearchTree) searchNode(root *TreeNode, key int) *TreeNode {
	// Se a árvore estiver vazia ou se o valor do nó atual for o valor procurado, retorna o nó
	if root == nil || root.data == key {
		return root
	}
	// Se o valor procurado for menor que o valor do nó atual, procura à esquerda
	if key < root.data {
		return bst.searchNode(root.left, key)
	}
	// Se o valor procurado for maior que o valor do nó atual, procura à direita
	return bst.searchNode(root.right, key)
}